### PR TITLE
Fix a bug for grad norm computation in SFT. Get loss 10% lower after fixing this bug

### PR DESCRIPTION
**Pull request by ZihanWang314**
Thanks for your great work on veRL! However, the current implementation in SFT for grad accumulation seems wrong. The loss should be divided first, then do backward. Experiments show that fixing this bug can get loss 10% lower (from 2.7 -> 2.4) on some tasks

### CODE REVIEW

File: verl/trainer/fsdp_sft_trainer.py [hunk 1]
```diff
@@ -394,8 +394,10 @@ def training_step(self, batch: TensorDict):
         n_micro_batches = len(micro_batches)
         step_loss = 0
         for micro_batch in micro_batches:
-            loss = self._compute_loss_and_backward(batch=micro_batch) / n_micro_batches
+            loss = self._compute_loss_and_backward(batch=micro_batch, do_backward=False) / n_micro_batches
+            loss.backward()
             step_loss += loss.item()
+            grad_norm = self.fsdp_model.clip_grad_norm_(max_norm=self.config.optim.clip_grad)
```
Comment (vermouth1992) [inline verl/trainer/fsdp_sft_trainer.py:400]: Is it correct to perform a global grad clip after all the micro batches forward and backward?

Comment (ZihanWang314) [reply verl/trainer/fsdp_sft_trainer.py:400]: Of course yes if the grad norm is divided first. Note that when every micro batch comes in, it only contributes to its fraction in the final batch, both for loss and for grad norm, so I'm very sure to implement correct micro-batching, each micro batch should be forward and backward first (with the loss divided before doing backward), then finally do the norm clip.

**Comment by ZihanWang314** [comment]
image
image
Please also see the loss curve and grad norm curve here. With current implementation, grad norm is very high when micro batch size is smaller than batch size

File: verl/trainer/fsdp_sft_trainer.py [hunk 2]
```diff
@@ -394,8 +394,10 @@ def training_step(self, batch: TensorDict):
         n_micro_batches = len(micro_batches)
         step_loss = 0
         for micro_batch in micro_batches:
-            loss = self._compute_loss_and_backward(batch=micro_batch) / n_micro_batches
+            loss = self._compute_loss_and_backward(batch=micro_batch, do_backward=False) / n_micro_batches
+            loss.backward()
```
Comment (xingyaoww) [inline verl/trainer/fsdp_sft_trainer.py:398]: I think we need to move the .backward() with in the context (https://github.com/volcengine/verl/pull/436/files#diff-07c65794985259af3f0f8e156fcf70119ad136cd42e06b4c6584799b1ebc8f2fR300)

otherwise SP will likely cause OOM issue

File: verl/trainer/fsdp_sft_trainer.py (hunk 1 above)
Comment (hiyouga) [reply verl/trainer/fsdp_sft_trainer.py:400]: It's ok to perform backprop for each microbatch data, but I think the `clip_grad_norm` operation can be done after computing gradients for all micro-batches.

Suggest change:
```python
        for micro_batch in micro_batches:
            loss = self._compute_loss_and_backward(batch=micro_batch, do_backward=False) / n_micro_batches
            loss.backward()
            step_loss += loss.item()

        grad_norm = self.fsdp_model.clip_grad_norm_(max_norm=self.config.optim.clip_grad)
```
